src/config/config.go

```diff
@@ -2,11 +2,36 @@ package config
 
 import (
 	"encoding/hex"
+	"sync"
 
 	"github.com/yggdrasil-network/yggdrasil-go/src/crypto"
 	"github.com/yggdrasil-network/yggdrasil-go/src/defaults"
 )
 
+// NodeState represents the active and previous configuration of the node and
+// protects it with a mutex
+type NodeState struct {
+	Current  NodeConfig
+	Previous NodeConfig
+	Mutex    sync.RWMutex
+}
+
+// Get returns both the current and previous node configs
+func (s *NodeState) Get() (NodeConfig, NodeConfig) {
+	s.Mutex.RLock()
+	defer s.Mutex.RUnlock()
+	return s.Current, s.Previous
+}
+
+// Replace the node configuration with new configuration
+func (s *NodeState) Replace(n NodeConfig) NodeConfig {
```

Comment:
If you call this and then call `Get()` to get the old config again, you potentially have a race. You could get around the race by returning the current and previous config when you call this (or just the previous, since the caller will already have the new config). This is (theoretically) an serious problem since `Replace` could be called again before all relevant goroutines have all called `Get` and made any necessary changes. In practice, this shouldn't normally be an issue, but it would be an absolute nightmare to debug if e.g. a node is under heavy load and gets another `SIGHUP` before it finishes updating everything.

Reply:
I've updated `Replace` to return both the current and previous configs now. I wasn't really sure of a better way to handle this.